aten/src/ATen/native/vulkan/api/Cache.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <ATen/native/vulkan/api/Common.h>
+
+namespace at {
+namespace native {
+namespace vulkan {
+namespace api {
+
+//
+// A generic cache for immutable Vulkan objects, when there will not be many
+// instances of those objects required at runtime.  The previous sentence puts
+// two constraints on proper use of this cache: 1) First, the objects should
+// preferably be immutable otherwise much care is required to synchronize
+// their usage.  2) Second, this cache is only intended for objects that
+// we will not have many instances of during the entire execution of the
+// program, otherwise the cache must be _infrequently_ purged.  Proper usage
+// model for this cache is in direct contrast with Vulkan object pools, which
+// indeed are required to be _frequently_ purged.  That is an important
+// distinction.
+//
+
+template<typename Factory>
+class Cache final {
+ public:
+  explicit Cache(Factory factory);
+  ~Cache() = default;
```

Comment:
Cache owns map that owns Handles with non trivial destructors(Deleters).
Should it have deleted copy-ctor and copy-assignment to prevent copy and double deletion of Handle's?

Reply:
Good point, and even better thinking!  I agree that defining the copy constructor and assignment operator as deleted would have been better practice and that's what I will do.

Having said that, there is no danger of double deletion.  That is because implicitly declared copy constructors (copy assignment operators) are defined as deleted if, among other conditions, at least one non-static data member of the class is not copyable (copy assignable).

From https://en.cppreference.com/w/cpp/language/copy_constructor:

> The implicitly-declared or defaulted copy constructor for class T is  defined as deleted if any of the following conditions are true:
> 
> -  T has non-static data members that cannot be copied (have deleted, inaccessible, or ambiguous copy constructors);
> -  T has direct or virtual base class that cannot be copied (has deleted, inaccessible, or ambiguous copy constructors);
> -  T has direct or virtual base class with a deleted or inaccessible destructor; 

That means all it takes for this class to end up having its implicitly declared copy constructor (copy assignment operator) marked as deleted is for either [one of its member variables] flat_hash_map or Factory to not be copyable (copy assignable), and flat_hash_map will in turn be non-copyable (non-copy assignable) if the data type it contains is not copyable (not copy-assignable), and so on.  Since at::native::vulkan::api::Handle has both copy constructor and assignment deleted, this whole chain will end up being non-copyable.

But I agree that making this explicit is better practice.  I will change the code accordingly.